File: src/main/java/org/pyjjs/scheduler/core/api/impl/SchedulerImpl.java
package org.pyjjs.scheduler.core.api.impl;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.actor.UntypedActor;
import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import org.jetbrains.annotations.NotNull;
import org.pyjjs.scheduler.core.api.impl.actors.resource.ResourceActor;
import org.pyjjs.scheduler.core.api.impl.actors.resource.supervisor.ResourceSupervisor;
import org.pyjjs.scheduler.core.api.impl.actors.system.ModificationController;
import org.pyjjs.scheduler.core.api.impl.actors.system.SchedulingController;
import org.pyjjs.scheduler.core.api.impl.actors.system.messages.*;
import org.pyjjs.scheduler.core.api.impl.actors.task.TaskActor;
import org.pyjjs.scheduler.core.api.impl.actors.task.supervisor.TaskSupervisor;
import org.pyjjs.scheduler.core.api.PlanMergingController;
import org.pyjjs.scheduler.core.api.PlanRepresentative;
import org.pyjjs.scheduler.core.api.Scheduler;
import org.pyjjs.scheduler.core.data.HashSetDataSourceImpl;
import org.pyjjs.scheduler.core.data.ObservableDataSource;
import org.pyjjs.scheduler.core.model.IdentifiableObject;
import org.pyjjs.scheduler.core.model.Resource;
import org.pyjjs.scheduler.core.model.Task;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nonnull;
import java.util.*;

import static com.google.common.base.Preconditions.checkNotNull;

public class SchedulerImpl implements Scheduler {

    private static final String DEFAULT_SCHEDULER_CONFIG_PATH = "default_scheduler_config.properties";

    private UUID uuid = UUID.randomUUID();

    private ObservableDataSource dataSource;

    private ActorSystem actorSystem;

    private Map<IdentifiableObject, ActorRef> agentToEntityMap = new HashMap<>();

    private BiMap<Class<? extends IdentifiableObject>, Class<? extends UntypedActor>> agentToEntityTypeMap = HashBiMap.create();

    private volatile boolean running = false;

    private ActorRef schedulingController;

    private ActorRef modificationController;

    private ActorRef taskSupervisor;

    private ActorRef resourceSupervisor;

    private PlanMergingController mergingController;

    private Logger LOG = LoggerFactory.getLogger(String.format("Scheduler [%s]", uuid.toString()));


    public SchedulerImpl() {
        this(new HashSetDataSourceImpl());
    }

    public SchedulerImpl(@Nonnull ObservableDataSource dataSource) {
        this.dataSource = dataSource;
        mergingController = new PlanMergingController();
        prepareAndLaunch();
    }

    private void prepareAndLaunch() {
        fillAgentToEntityClassMapping();
        Config config = ConfigFactory.load(DEFAULT_SCHEDULER_CONFIG_PATH);
        actorSystem = ActorSystem.apply("scheduler", config);
    }

    private void createDataSourceModificationControllerListener() {
        List<Class<? extends IdentifiableObject>> interests = Arrays.asList(Task.class, Resource.class);
        this.dataSource.addDataSourceListener(interests, new ObservableDataSource.DataSourceListener() {

            @Override
            public void onCreate(IdentifiableObject entity) {
                EntityCreatedMessage msg = new EntityCreatedMessage(entity);
                modificationController.tell(msg, ActorRef.noSender());
            }

            @Override
            public void onUpdate(IdentifiableObject entity) {
                EntityUpdatedMessage msg = new EntityUpdatedMessage(entity);
                modificationController.tell(msg, ActorRef.noSender());
            }

            @Override
            public void onRemove(IdentifiableObject entity) {
                EntityRemovedMessage msg = new EntityRemovedMessage(entity);
                modificationController.tell(msg, ActorRef.noSender());
            }
        });
    }

    private void createSystemAgents() {
        taskSupervisor = actorSystem.actorOf(Props.create(TaskSupervisor.class), "tasks");
        resourceSupervisor = actorSystem.actorOf(Props.create(ResourceSupervisor.class), "resources");

        schedulingController = actorSystem.actorOf(Props.create(SchedulingController.class, mergingController));
        actorSystem.eventStream().subscribe(schedulingController, PlanUpdatedMessage.class);

        modificationController = actorSystem.actorOf(Props.create(ModificationController.class, taskSupervisor, resourceSupervisor));
    }

    private BiMap<Class<? extends IdentifiableObject>, Class<? extends UntypedActor>> fillAgentToEntityClassMapping() {
        agentToEntityTypeMap.put(Resource.class, ResourceActor.class);
        agentToEntityTypeMap.put(Task.class, TaskActor.class);
        return agentToEntityTypeMap;
    }

    @Override
    public void run() {
        initAndStart();
    }

    private void initAndStart() {
        LOG.info("Scheduling init started");
        removeExistedActors();
        createSystemAgents();
        notifyModificationControllerAboutExistedEntities();
        createDataSourceModificationControllerListener();
    }

    private void notifyModificationControllerAboutExistedEntities() {
        for (IdentifiableObject identifiableObject : dataSource) {
            EntityCreatedMessage entityCreatedMessage = new EntityCreatedMessage(identifiableObject);
            modificationController.tell(entityCreatedMessage, ActorRef.noSender());
        }
    }

    private Map<IdentifiableObject, ActorRef> fillActorMapByDataSource() {
        for (IdentifiableObject entity : dataSource) {
            Class<? extends UntypedActor> actorType = agentToEntityTypeMap.get(entity.getClass());
            checkNotNull(actorType, "In data source found entity not extended Resource or Task");
            ActorRef entityActor = createActorByType(actorSystem, actorType);
            agentToEntityMap.put(entity, entityActor);
        }
        return agentToEntityMap;
    }

    private static ActorRef createActorByType(ActorSystem actorSystem, Class<? extends UntypedActor> actorType) {
        return actorSystem.actorOf(Props.create(actorType));
    }

    @Override
    public boolean isRunning() {
        return running;
    }

    @Override
    public void reset() {
        removeExistedActors();
    }

    private void removeExistedActors() {
        for (ActorRef actorRef : agentToEntityMap.values()) {
            actorSystem.stop(actorRef);
        }
        agentToEntityMap.clear();

        if(modificationController != null) {
            actorSystem.stop(modificationController);
            modificationController = null;
        }

        if(taskSupervisor != null) {
            actorSystem.stop(taskSupervisor);
            taskSupervisor = null;
        }

        if(resourceSupervisor != null) {
            actorSystem.stop(resourceSupervisor);
            resourceSupervisor = null;
        }

        if(schedulingController != null) {
            actorSystem.stop(schedulingController);
            schedulingController = null;
        }
    }

    @NotNull
    @Override
    public ObservableDataSource getDataSource() {
        return dataSource;
    }

    @Override
    public void addStablePlanListener(@Nonnull PlanRepresentative.StablePlanListener listener) {
        mergingController.addStablePlanListener(listener);
    }

    @Override
    public void removeStablePlanListener(@Nonnull PlanRepresentative.StablePlanListener listener) {
        mergingController.removeStablePlanListener(listener);
    }
}

File: src/main/kotlin/org/pyjjs/scheduler/core/api/impl/SchedulerImpl.kt
package org.pyjjs.scheduler.core.api.impl

import akka.actor.ActorRef
import akka.actor.ActorSystem
import akka.actor.Props
import akka.actor.UntypedActor
import akka.japi.Creator
import com.google.common.base.Preconditions.checkNotNull
import com.google.common.collect.BiMap
import com.google.common.collect.HashBiMap
import com.typesafe.config.ConfigFactory
import org.pyjjs.scheduler.core.api.PlanMergingController
import org.pyjjs.scheduler.core.api.PlanRepresentative
import org.pyjjs.scheduler.core.api.Scheduler
import org.pyjjs.scheduler.core.api.impl.actors.common.messages.TaskAppearedMessage
import org.pyjjs.scheduler.core.api.impl.actors.resource.ResourceActor
import org.pyjjs.scheduler.core.api.impl.actors.resource.supervisor.ResourceSupervisor
import org.pyjjs.scheduler.core.api.impl.actors.system.ModificationController
import org.pyjjs.scheduler.core.api.impl.actors.system.SchedulingController
import org.pyjjs.scheduler.core.api.impl.actors.system.messages.EntityCreatedMessage
import org.pyjjs.scheduler.core.api.impl.actors.system.messages.EntityRemovedMessage
import org.pyjjs.scheduler.core.api.impl.actors.system.messages.EntityUpdatedMessage
import org.pyjjs.scheduler.core.api.impl.actors.system.messages.PlanUpdatedMessage
import org.pyjjs.scheduler.core.api.impl.actors.task.TaskActor
import org.pyjjs.scheduler.core.api.impl.actors.task.supervisor.TaskSupervisor
import org.pyjjs.scheduler.core.data.HashSetDataSourceImpl
import org.pyjjs.scheduler.core.data.ObservableDataSource
import org.pyjjs.scheduler.core.model.IdentifiableObject
import org.pyjjs.scheduler.core.model.Resource
import org.pyjjs.scheduler.core.model.Task
import org.slf4j.LoggerFactory
import java.util.*

class SchedulerImpl @JvmOverloads constructor(override val dataSource: ObservableDataSource = HashSetDataSourceImpl()) : Scheduler {

    private val uuid = UUID.randomUUID()

    private lateinit var actorSystem: ActorSystem

    private val agentToEntityMap = HashMap<IdentifiableObject, ActorRef>()

    private val agentToEntityTypeMap = HashBiMap.create<Class<out IdentifiableObject>, Class<out UntypedActor>>()

    @Volatile override var isRunning = false

    private var schedulingController: ActorRef? = null

    private var modificationController: ActorRef? = null

    private var taskSupervisor: ActorRef? = null

    private var resourceSupervisor: ActorRef? = null

    private val mergingController: PlanMergingController

    private val LOG = LoggerFactory.getLogger(String.format("Scheduler [%s]", uuid.toString()))

    init {
        mergingController = PlanMergingController()
        prepareAndLaunch()
    }

    private fun prepareAndLaunch() {
        fillAgentToEntityClassMapping()
        val config = ConfigFactory.load("default_scheduler_config.properties")
        actorSystem = ActorSystem.apply("scheduler", config)
    }

    private fun createDataSourceModificationControllerListener() {
        val interests = Arrays.asList(Task::class.java, Resource::class.java)
        this.dataSource.addDataSourceListener(interests, object : ObservableDataSource.DataSourceListener {

            override fun onCreate(entity: IdentifiableObject) {
                val msg = EntityCreatedMessage(entity)
                modificationController!!.tell(msg, ActorRef.noSender())
            }

            override fun onUpdate(entity: IdentifiableObject) {
                val msg = EntityUpdatedMessage(entity)
                modificationController!!.tell(msg, ActorRef.noSender())
            }

            override fun onRemove(entity: IdentifiableObject) {
                val msg = EntityRemovedMessage(entity)
                modificationController!!.tell(msg, ActorRef.noSender())
            }
        })
    }

    private fun createSystemAgents() {
        taskSupervisor = actorSystem.actorOf(Props.create(TaskSupervisor::class.java), "tasks")
        resourceSupervisor = actorSystem.actorOf(Props.create(ResourceSupervisor::class.java), "resources")

        schedulingController = actorSystem.actorOf(Props.create(SchedulingController::class.java, mergingController))
        actorSystem.eventStream().subscribe(schedulingController, PlanUpdatedMessage::class.java)
        actorSystem.eventStream().subscribe(schedulingController, TaskAppearedMessage::class.java)

        modificationController = actorSystem.actorOf(Props.create(ModificationController::class.java, taskSupervisor, resourceSupervisor))
    }

    private fun fillAgentToEntityClassMapping(): BiMap<Class<out IdentifiableObject>, Class<out UntypedActor>> {
        agentToEntityTypeMap.put(Resource::class.java, ResourceActor::class.java)
        agentToEntityTypeMap.put(Task::class.java, TaskActor::class.java)
        return agentToEntityTypeMap
    }

    override fun run() {
        initAndStart()
    }

    private fun initAndStart() {
        LOG.info("Scheduling init started")
        removeExistedActors()
        createSystemAgents()
        notifyModificationControllerAboutExistedEntities()
        createDataSourceModificationControllerListener()
    }

    private fun notifyModificationControllerAboutExistedEntities() {
        for (identifiableObject in dataSource) {
            val entityCreatedMessage = EntityCreatedMessage(identifiableObject)
            modificationController!!.tell(entityCreatedMessage, ActorRef.noSender())
        }
    }

    override fun reset() {
        removeExistedActors()
    }

    private fun removeExistedActors() {
        for (actorRef in agentToEntityMap.values) {
            actorSystem.stop(actorRef)
        }
        agentToEntityMap.clear()

        if (modificationController != null) {
            actorSystem.stop(modificationController)
            modificationController = null
        }

        if (taskSupervisor != null) {
            actorSystem.stop(taskSupervisor)
            taskSupervisor = null
        }

        if (resourceSupervisor != null) {
            actorSystem.stop(resourceSupervisor)
            resourceSupervisor = null
        }

        if (schedulingController != null) {
            actorSystem.stop(schedulingController)
            schedulingController = null
        }
    }

    override fun addStablePlanListener(listener: PlanRepresentative.StablePlanListener) {
        mergingController.addStablePlanListener(listener)
    }

    override fun removeStablePlanListener(listener: PlanRepresentative.StablePlanListener) {
        mergingController.removeStablePlanListener(listener)
    }
}
